extensions/async-extensions/src/commonMain/kotlin/app/cash/sqldelight/async/coroutines/QueryExtensions.kt

```diff
@@ -4,10 +4,20 @@ import app.cash.sqldelight.ExecutableQuery
 import app.cash.sqldelight.db.QueryResult
 
 suspend fun <T : Any> ExecutableQuery<T>.awaitAsList(): List<T> = execute { cursor ->
-  QueryResult.AsyncValue {
-    val result = mutableListOf<T>()
-    while (cursor.next().await()) result.add(mapper(cursor))
-    result
+  val first = cursor.next()
+  val result = mutableListOf<T>()
+
+  // If the cursor isn't async, we want to preserve the blocking semantics and execute it synchronously
+  if (first is QueryResult.AsyncValue) {
+    QueryResult.AsyncValue {
+      if (first.await()) result.add(mapper(cursor)) else return@AsyncValue result
+      while (cursor.next().value) result.add(mapper(cursor))
+      result
+    }
+  } else {
+    if (first.value) result.add(mapper(cursor)) else return@execute QueryResult.Value(result)
+    while (cursor.next().value) result.add(mapper(cursor))
+    QueryResult.Value(result)
```

Comment:
This change deserves extra attention because of how... special it is.

Reply:
merged cause green, I'm surprised this was needed, I assumed calling `await` on a `QueryResult.Value` would just block anyway and do the same thing as `.value`

Reply:
Explaining it here for posterity:

`AsyncValue`s are evaluated lazily so that the `getter` block can be run in the consumer's coroutine scope. Because of this, when `awaitAsList()` (or `awaitAsOne...()`) is called, the actual work in the `AsyncValue` block is deferred until the final `.await()` is called, which happens after the call to `execute()` has completed. So by then, the driver may have closed the connection, and there will be no results in the cursor.

Basically the order of calls before:
1. `awaitAs*()`
2. `execute()`
3. `mapper()`
4. `cursor.next()` _(synchronously called in the mapper)_
5. `onClose()` _(driver closes the connection)_
6. `await()` _(returns the synchronous value)_

and after:
1. `awaitAs*()`
2. `execute()`
3. `mapper()` _(returns an `AsyncValue` which will be lazily evaluated)_
4. `onClose()` _(driver closes the connection)_
5. `await()` _(evaluates the `AsyncValue`)_
6. `cursor.next()` _(connection has already been closed)_